### DISCUSSION TITLE
[Proposal] Allow scope resolution operator on class name to define nested types

### DISCUSSION THREAD

**Discussion by muhamad** [dotnet/csharplang#1894] [Language Ideas]
This proposal is to define a syntax to allow creating nested types of a class in a more friendly way so instead of:
```csharp
public class A
{
    private interface IB {}
    private class B : IB {}
}
```
we use the following syntax
```csharp
public partial class A {}

private interface A::IB {}

private class A::B : A.IB {}
```
and if we want to define the nested class `B` to be partial is it OK using the current existing syntax:
```csharp
// in file AB_1.cs
private partial class A::B : A.IB {}

// in file AB_2.cs
private partial class A::B : A.IB {}
```
the scope resolution operator here allowed moving the nested `class\interface\struct\enum` out of the main class definition under one condition that the class `A` to be `partial` so when the compiler merge the proposed code we get the original definition.

the new syntax will allow simplifying the nested class type definition to keep the code for large classes clean and maintainable.

**Comment by HaloFour** [comment]
This is all already possible:

```cs
// in file AB_1.cs
public partial class A {
    private interface IB { }
}

// in file AB_2.cs
public partial class A {
    private class B : IB { }
}
```

I don't see why you'd want to move the declaration out from the parent `A` class.  That relationship is very important since the nested types have privileged access to the members of the parent type.

The `::` operator also [already exists](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/namespace-alias-qualifer) in C#.

**Comment by Austin-bryan** [comment]
Also, the `::` operator means something else in C++ and is used in a similar situation, so it would be quite confusing. Why not use `A.B` or something as that's already the way C# uses to describe that relationship, like if `B` was public, and you're writing outside of `A`, you can do, `var b = new A.B();`.

Regardless this sounds an awful lot like the idea to write local functions outside of the parent function and I imagine it would get rejected for the same reason, it making *no sense*. 

If you're really worried that you're one class is too long because it's hundreds of lines, but that's just because it has child classes and the main class is only 100 lines, then you're still likely following good practices it's fine. 

And as @HaloFour already pointed out, you can just use partial classes.

**Comment by dsaf** [comment]
If there will be unnested nested classes then there should be named anonymous classes as well!

**Comment by Austin-bryan** [comment]
@dsaf And global local variables!
